Introducing block strings

vitalibarozzi opened the pull request:
### Motivation

Current implementation is missing **Block Strings,** as mentioned
in the spec. In here we describe what changed, how and why.

Related to #30.

### Reasoning

Since we needed an implementation that preserved block strings
in the AST so we could reprint it as a block string and not
just as normal string, we opted for adding a new constructor
to the AST. For this constructor we added a new parser, printer
and generator, besides the needed tests.

In the old tests we changed the property tests for the cycle
_"parse→print→parse"_ so they can have a notion of idempotency
that also accommodates the block strings; the thing with block strings
is that after parsed once the excess indentation is removed
and when printed the resulting string is different (although
equivalent) to the original one, so the property _"parse (print
ast) = ast"_ is not true in the first parsing, but it will be
true to all the subsequent ones (where the common indentation
stays at zero forever).

### Implementation

In the generator file now we have the main generator for
block strings _"genBlockText"_ and a couple helper functions. In
the parser file we changed the value parser to include block
strings, and we also created the _"blockString"_ parser. The
printer is pretty much standard and in the syntax file we just
added the constructor and updated the needed instances.

We added the unit tests for block strings and the idempotency test.
The idempotency is tested both in the old property tests that now
check the idempotency in one or two parsings, but we also test
it in the file where the future property tests for block strings
will be.

### Consequences

Probably not many; the changes should impact normal strings,
descriptions and block strings. Nothing else should be affected.

### Further work to be done

We decided to first get the input of you guys in what has been
done so far so we can proceed with the missing stuff, which are:

- Optimizations. We have a naive implementation.
- Some property tests are just placeholders at the moment.
- Unit tests might be turned into property tests. (at least a couple)

lorenzo commented:
@codingkarthik Would you mind taking a look at this? We are not too happy with how we had to change the test assertions, do you have an idea on how to improve this?

codingkarthik commented on src/Language/GraphQL/Draft/Syntax.hs:
```diff
@@ -320,6 +320,7 @@ data Value var
   | VInt Integer
   | VFloat Scientific
   | VString Text
+  | VBlockString Text
```
I was thinking instead of adding a new constructor `VBlockString` if we could extend the `VString` type in the following manner.

```haskell
data StringType = StringCharacter | BlockStringCharacter

data Value var
  = VVariable var
  | VNull
  | VInt Integer
  | VFloat Scientific
  | VString StringType Text
  | VBoolean Bool
  | VEnum EnumValue
  | VList [Value var]
  | VObject (HashMap Name (Value var))
```

This way both the string types are clubbed together and I think except the parsing and printing bit, both types of string will be processed in the same way.

codingkarthik commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +489,40 @@ between open close p = tok open *> p <* tok close
 optempty :: Monoid a => Parser a -> Parser a
 optempty = option mempty
+
+-- | Parses strings delimited by triple quotes.
+blockString :: Parser Text
```
```suggestion
-- | Parses strings delimited by triple quotes. 
--   http://spec.graphql.org/June2018/#sec-String-Value
blockString :: Parser Text
```

codingkarthik reviewed (changes requested):
Thank you so much for opening this PR! @lorenzo @vitalibarozzi 
Mostly the changes LGTM, I have left couple of comments in line.

I'm not verse with the testing part here but I'll make sure someone who is to review it!

evertedsphere commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +489,40 @@ between open close p = tok open *> p <* tok close
+  return (T.strip . toStrict . TB.toLazyText $ fixedLines)
+ where
+  -- used to remove the common indentation from each line.
+  fixIdentation :: Int -> Text -> TB.Builder -> TB.Builder 
```
```suggestion
  fixIndentation :: Int -> Text -> TB.Builder -> TB.Builder
```
and ditto below

lorenzo commented:
@codingkarthik @evertedsphere thanks for the speedy review!

Would you mind giving it a second look?

abooij commented on src/Language/GraphQL/Draft/Generator.hs:
```diff
@@ -101,6 +103,43 @@ genObjectValue genVal = M.fromList <$> mkList genObjectField
+genBlockText =
+  Gen.choice [ simple, genLines ]
+ where
+  simple = do
+    Gen.frequency
+      [ (10, Gen.text (Range.linear 1 100) Gen.unicode)
+      , (10, return "\n")
+      , (6, genIndentation)
+      , (5, genMinIndentedText 10)
+      , (4, return "")
+      , (3, return " ")
+      , (3, return "\"") -- "
+      , (3, return "\\") -- \
+      ]
+  genLines :: Gen Text
+  genLines = do
+    n <- Gen.int (Range.linear 1 100)
+    x <- Gen.list (Range.linear 1 n) simple
+    return (T.unlines x)
```
This looks to me like it's only testing a non-zero number of lines, each of non-zero length - is that right? Can we include some `0`s in these options?

abooij commented on src/Language/GraphQL/Draft/Generator.hs:
```diff
@@ -101,6 +103,43 @@ genObjectValue genVal = M.fromList <$> mkList genObjectField
+  t <- genText
+  return (minIndent <> i <> t)
+
+genIndentation :: Gen Text
+genIndentation = do
+  n <- Gen.int (Range.linear 1 30)
+  T.concat <$> Gen.choice
+    [ Gen.list (Range.linear 1 n) (return " ")
+    , Gen.list (Range.linear 1 (n * 100)) (return " ")
+    ]
```
Ditto, does this include having no indentation?

abooij commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +489,41 @@ between open close p = tok open *> p <* tok close
+blockString = do
+  _ <- triplequotes <?> "opening triple quotes"
+  str <- T.lines . T.pack <$> AT.manyTill AT.anyChar triplequotes <?> "the body of a triple quoted string"
+  let !cleanLines = dropEmptyLines str
```
What does this bang do? It certainly won't force the entire list to normal form.

abooij commented on test/BlockStrings.hs:
```diff
@@ -0,0 +1,42 @@
+    [ ("parses the specExample", blockParsesTo "\n    Hello,\n      World!\n\n    Yours,\n      GraphQL.\n  " "Hello,\n  World!\n\nYours,\n  GraphQL.")
+    , ("parses empty string", blockParsesTo "" "")
+    , ("parses newline", blockParsesTo "\n" "")
+    , ("parses newlines in empty block", blockParsesTo " \n   \n  \n            \n\n " "")
```
I don't think this is what the [specification](http://spec.graphql.org/June2018/#BlockStringValue()) says. I think `BlockStringValue("\n \n \n \n\n ")` would compute to be `" \n \n \n"`.

vitalibarozzi replied:
So we use the indentation of the last line to determine the smallest indentation, which is 0, and so do not remove the indentation of the previous lines even though the last line will be removed in the end of the reformatting? Is that the correct understanding? This is a weird one.

abooij replied:
Looking at this again, I'm not sure I interpreted the spec correctly. Perhaps this test is correct after all. `BlockStringValue` is an odd beast.

vitalibarozzi replied:
Yeah, I'm having some trouble with this one, but you are correct about removing only first and last lines anyway.

vitalibarozzi replied:
We both got confused.
As Lorenzo pointed out to me in the spec for BlockStringValue we remove all empty lines case there is only empty lines, because the step 5 includes a **while** that we missed.
So I going to guide the implementation in this direction.

abooij commented on test/BlockStrings.hs:
```diff
@@ -0,0 +1,42 @@
+    , ("parses newline", blockParsesTo "\n" "")
+    , ("parses newlines in empty block", blockParsesTo " \n   \n  \n            \n\n " "")
+    , ("parses very simples not-empty block", blockParsesTo "x" "x")
+    , ("parses block string containing a valid normal string inside", blockParsesTo "  \"i'm like a JSON string\"   " "\"i'm like a JSON string\"")
```
I think `BlockStringValue(" \"i'm like a JSON string\"")` computes to `" \"i'm like a JSON string\""` (no indentation removed since it's on the first line).

abooij commented on test/BlockStrings.hs:
```diff
@@ -0,0 +1,42 @@
+    , ("parses newlines in empty block", blockParsesTo " \n   \n  \n            \n\n " "")
+    , ("parses very simples not-empty block", blockParsesTo "x" "x")
+    , ("parses block string containing a valid normal string inside", blockParsesTo "  \"i'm like a JSON string\"   " "\"i'm like a JSON string\"")
+    , ("parses not-empty block with newlines", blockParsesTo " \n   \n  \n x   \n y   \n\n " "x\ny")
```
I don't think `BlockStringValue` removes whitespace at the end of the line.

abooij commented on test/BlockStrings.hs:
```diff
@@ -0,0 +1,42 @@
+    , ("parses very simples not-empty block", blockParsesTo "x" "x")
+    , ("parses block string containing a valid normal string inside", blockParsesTo "  \"i'm like a JSON string\"   " "\"i'm like a JSON string\"")
+    , ("parses not-empty block with newlines", blockParsesTo " \n   \n  \n x   \n y   \n\n " "x\ny")
+    , ("ignores escaping", blockParsesTo "  \\  " "\\") -- this is a single \
```
I don't think whitespace is removed from the first line.

abooij commented on test/BlockStrings.hs:
```diff
@@ -0,0 +1,42 @@
+    , ("parses block string containing a valid normal string inside", blockParsesTo "  \"i'm like a JSON string\"   " "\"i'm like a JSON string\"")
+    , ("parses not-empty block with newlines", blockParsesTo " \n   \n  \n x   \n y   \n\n " "x\ny")
+    , ("ignores escaping", blockParsesTo "  \\  " "\\") -- this is a single \
+    , ("\n in first characters is parsed", blockParsesTo "\n hey  " "hey")
```
I don't think whitespace at the end of the line is removed.

abooij commented on test/BlockStrings.hs:
```diff
@@ -0,0 +1,42 @@
+    , ("parses not-empty block with newlines", blockParsesTo " \n   \n  \n x   \n y   \n\n " "x\ny")
+    , ("ignores escaping", blockParsesTo "  \\  " "\\") -- this is a single \
+    , ("\n in first characters is parsed", blockParsesTo "\n hey  " "hey")
+    , ("zero common indentation is possible", blockParsesTo " \naa \n  bbbb \ncccc " "aa\n  bbbb\ncccc")
+    , ("common indentation is removed", blockParsesTo " \n   i have 3 \n    i have 4 \n   i also have 3 " "i have 3\n i have 4\ni also have 3")
```
ditto.

abooij commented on test/Spec.hs:
```diff
@@ -74,13 +76,20 @@ propParserBSPrinter = mkPropParserPrinter $ bsToTxt . BS.toLazyByteString . Outp
-    xs <- forAll genExecutableDocument
-    let rendered = printer xs
-    either onError (xs ===) $ Input.parseExecutableDoc rendered
+    someRandomDoc <- forAll genExecutableDocument
+    let rendered = printer <$> Input.parseExecutableDoc (printer someRandomDoc)
+    -- the reason why we render twice is to make sure that
+    -- when comparing we already "normalized" the block
+    -- string, so each reparsing of the blockstring will no
+    -- longer be different because the indentations are
+    -- already at the minimun possible.
+    let reRendered = printer <$> (Input.parseExecutableDoc =<< rendered)
+    case (rendered, reRendered) of
+        (Left  e, _      ) -> onError e
+        (_      , Left  e) -> onError e
+        (Right a, Right b) -> a === b
```
Does this mean that this PR weakens the test suite?

lorenzo replied:
unfortunately. I’m really looking forward to finding a better way. Anything by comes to mind?

abooij replied:
In light of the recent discovery (#44) that the test suite of this library was inadequate, I think it's really important that we figure this one out.

abooij reviewed (commented):
Thanks a lot for this work. It's indeed about time to add support for this syntax!

I don't really understand the motivation for having either an additional `VBlockString` constructor, or for adding an additional field to `VString`. The "Reasoning" section above suggests that you want to be able to print block strings sometimes. Why don't we just make the pretty printer choose to print a block string whenever there are newlines in the strings, and regular strings otherwise? That way we don't need to change the definition of `Value`, and it also avoids having descriptions with several paragraphs printed as non-block strings (which admittedly would not look good).

abooij commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -198,27 +205,30 @@ booleanLiteral
-    -- reconstructing a literal string (by putting quotes around it) and
-    -- delegating all the hard work to Aeson.
-    unescapeText str = either fail pure $ A.parseOnly jstring ("\"" <> T.encodeUtf8 str <> "\"")
+stringLiteral = 
+  unescapeText =<< (char '"' *> jstring_ <?> "string")
+ where
+  -- | Parse a string without a leading quote, ignoring any escaped characters.
+  jstring_ :: Parser Text
+  jstring_ = scan False go <* anyChar
+
+  go :: Bool -> Char -> Maybe Bool
+  go previousWasEscapingCharacter current
+    -- if the previous character was an escaping character, we skip this one
+    | previousWasEscapingCharacter = Just False
+    -- otherwise, if we find an unescaped quote, we've reached the end
+    | current == '"' = Nothing
+    -- otherwise, we continue, and track whether the current character is an escaping backslash
+    | otherwise = Just $ current == backslash
+    where backslash = '\\'
+
+  -- | Unescape a string.
+  --
+  -- Turns out this is really tricky, so we're going to cheat by
+  -- reconstructing a literal string (by putting quotes around it) and
+  -- delegating all the hard work to Aeson.
+  unescapeText :: Text -> Parser Text
+  unescapeText str = either fail pure $ A.parseOnly jstring ("\"" <> T.encodeUtf8 str <> "\"")
```
This new version has different indentation style from the rest of the library.

abooij commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +489,41 @@ between open close p = tok open *> p <* tok close
+  str <- T.lines . T.pack <$> AT.manyTill AT.anyChar triplequotes <?> "the body of a triple quoted string"
+  let !cleanLines = dropEmptyLines str
+  let smallest = foldr selectSmallest maxBound cleanLines -- counts indentation and selects the smallest one from all lines
+  let fixedLines = foldr (fixIndentation smallest) "" cleanLines -- removes the common indentation from all lines.
+  return (T.strip . toStrict . TB.toLazyText $ fixedLines)
+ where
+  -- used to remove the common indentation from each line.
+  fixIndentation :: Int -> Text -> TB.Builder -> TB.Builder 
+  fixIndentation smallest a acc =
```
Ditto, I think it would be cleaner to stick to the existing indentation style.

abooij replied:
```suggestion
  let fixedLines = foldr (fixIndentation smallest) "" cleanLines -- removes the common indentation from all lines.
  return (T.strip . toStrict . TB.toLazyText $ fixedLines)
 where
  -- used to remove the common indentation from each line.
  dropIndentation :: Int -> Text -> TB.Builder -> TB.Builder 
  dropIndentation smallest a acc =
```

abooij replied:
I don't think my first comment above has been resolved.

lorenzo replied:
I'll take a look at it

abooij commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +489,41 @@ between open close p = tok open *> p <* tok close
+  countIndentation = fromMaybe 0 . T.findIndex (/= ' ')
+  triplequotes = AT.string "\"\"\""
+
+onlySpaces :: Text -> Bool
+onlySpaces = T.all (== ' ')
+
+dropEmptyLines :: [Text] -> [Text]
+dropEmptyLines =  
+  dropWhileEnd cond . dropWhile cond
+ where
+  cond l = T.null l || onlySpaces l
+  dropWhileEnd p = foldr (\x xs -> if p x && null xs then [] else x : xs) []
```
What about tabs?

abooij replied:
Is there a way we could make the last line (with the `foldr`) easier to understand?

abooij commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +489,41 @@ between open close p = tok open *> p <* tok close
+       then new
+       else acc
+  -- used to count indentation in a single a line
+  countIndentation :: Text -> Int
+  countIndentation = fromMaybe 0 . T.findIndex (/= ' ')
```
What about tabs?

abooij commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +489,41 @@ between open close p = tok open *> p <* tok close
+    let new = TB.fromText $ T.dropWhileEnd (== ' ') (T.drop smallest a)
+    in new <> "\n" <> acc
+  -- used to accumulate the smallest indentation
+  selectSmallest :: Text -> Int -> Int
+  selectSmallest t acc =
+    let new = countIndentation t
+     in if (not $ onlySpaces t) && new < acc
+       then new
+       else acc
```
Instead of folding over the lines like this, why don't we just take a `min` over the `countIndentation`s?

vitalibarozzi replied:
On a first attempt it did not work. I'll try again tomorrow.

abooij replied:
I should have spelled this out more precisely. What I meant was that the `foldr` (which is difficult to read) is replaced entirely by taking the `minimum` over some appropriate list. Something along these lines:
```haskell
let commonIndentation = minimum $ (maxBound:) $ countIndentation <$> tail_
```

vitalibarozzi commented:
> Thanks a lot for this work. It's indeed about time to add support for this syntax!
> 
> I don't really understand the motivation for having either an additional `VBlockString` constructor, or for adding an additional field to `VString`. The "Reasoning" section above suggests that you want to be able to print block strings sometimes. Why don't we just make the pretty printer choose to print a block string whenever there are newlines in the strings, and regular strings otherwise? That way we don't need to change the definition of `Value`, and it also avoids having descriptions with several paragraphs printed as non-block strings (which admittedly would not look good).

Great idea.

abooij commented on src/Language/GraphQL/Draft/Printer.hs:
```diff
@@ -236,20 +236,27 @@ variableP v = charP '$' <> printP v
 stringValue :: Printer a => Text -> a
 stringValue s = textP $ LT.toStrict $ LTE.decodeUtf8 $ J.encode s
 
+blockStringValue :: Printer a => Text -> a
+blockStringValue t = textP "\"\"\"\n" <> textP t <> textP "\n\"\"\""
+
```
This fails in the same way that `stringValue` used to before it was fixed in #44.

lorenzo replied:
That means we cannot use the same constructor for both normal strings and block strings, or Am I missing something here?

vitalibarozzi replied:
It's not failing anymore with the new parser.

vitalibarozzi commented:
We just pushed a new version for the parser. Could you give it a check @abooij ? Thanks. :)

abooij commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +488,41 @@ between open close p = tok open *> p <* tok close
 optempty :: Monoid a => Parser a -> Parser a
 optempty = option mempty
+
+-- | Parses strings delimited by triple quotes. 
+-- http://spec.graphql.org/June2018/#sec-String-Value
+blockString :: Parser Text
+blockString = do
+  _ <- tripleQuotes <?> "opening triple quotes"
+  lines_ <- T.lines . T.pack <$> AT.manyTill AT.anyChar tripleQuotes <?> "the body of a triple quoted string"
+  let headline = if lines_ == [] then "" else head lines_
+  let tail_ = drop 1 lines_ -- not tail
+  let commonIndentation = foldr min maxBound (countIndentation <$> tail_)
+  let rlines = foldr (removeCommonIndentation commonIndentation) Nothing tail_
+  return $ case rlines of
+    Nothing -> headline
+    Just reformatedLines -> rebuild (sanitize $ headline:reformatedLines)
+ where
+  rebuild :: [Text] -> Text
+  rebuild = fromMaybe "" . fmap fst . T.unsnoc . T.unlines
+  sanitize :: [Text] -> [Text]
+  sanitize = dropWhileEnd' onlyWhiteSpace  . dropWhile onlyWhiteSpace
+  onlyWhiteSpace :: Text -> Bool
+  onlyWhiteSpace t = T.null t || T.all ws t
+  removeCommonIndentation :: Int -> Text -> Maybe [Text] -> Maybe [Text] 
+  removeCommonIndentation smallest a x = 
+    let new = T.drop smallest a
+     in case x of
+      Nothing -> Just [new]
+      Just acc -> Just $ new:acc
+  countIndentation :: Text -> Int
+  countIndentation = fromMaybe maxBound . T.findIndex (not . ws)
+  tripleQuotes = AT.string "\"\"\""
```
There seems to be a minor indentation inconsistency in, perhaps ironically, `removeCommonIndentation`.

abooij commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +488,41 @@ between open close p = tok open *> p <* tok close
+  rebuild = fromMaybe "" . fmap fst . T.unsnoc . T.unlines
+  sanitize :: [Text] -> [Text]
+  sanitize = dropWhileEnd' onlyWhiteSpace  . dropWhile onlyWhiteSpace
+  onlyWhiteSpace :: Text -> Bool
+  onlyWhiteSpace t = T.null t || T.all ws t
```
```suggestion
  onlyWhiteSpace :: Text -> Bool
  onlyWhiteSpace t = T.all ws t
```

abooij commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +488,41 @@ between open close p = tok open *> p <* tok close
+blockString = do
+  _ <- tripleQuotes <?> "opening triple quotes"
+  lines_ <- T.lines . T.pack <$> AT.manyTill AT.anyChar tripleQuotes <?> "the body of a triple quoted string"
+  let headline = if lines_ == [] then "" else head lines_
+  let tail_ = drop 1 lines_ -- not tail
```
For some reason these two lines seem quite sketchy to me. As a start, we could do them simultaneously with a pattern match. But, more worryingly, this might be introducing some extra edge cases.
What's the role of `headline`? I suppose it's got something to do with the special treatment of the first line by `BlockStringValue` - if so, let's make this explicit, at least with a comment.
`tail_` is not a very helpful variable name. Is this the possibly-indented remainder of the block string? Then we could call it `indentedRemainder`.

codingkarthik replied:
> Is this the possibly-indented remainder of the block string?

I think it is the non-indented remainder of the block string on which the common indentation is to be removed, point 4 [here](http://spec.graphql.org/June2018/#BlockStringValue())

abooij commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +488,41 @@ between open close p = tok open *> p <* tok close
+ws :: Char -> Bool
+ws c = c == ' ' || c == '\t'
+
+-- copied from https://hackage.haskell.org/package/extra-1.7.9/docs/src/Data.List.Extra.html
+dropWhileEnd' :: (a -> Bool) -> [a] -> [a]
+dropWhileEnd' p = foldr (\x xs -> if null xs && p x then [] else x : xs) []
```
There's a minor question of copyright we have to figure out before this PR is merged. Please don't resolve this comment until we figure out if this is OK.

abooij commented on test/BlockStrings.hs:
```diff
@@ -0,0 +1,44 @@
+    , ("parses very simples not-empty block", blockParsesTo "x" "x")
+    , ("common indentation is removed", blockParsesTo "\n  a \n   b \n  c " "a \n b \nc ")
+    , ("zero common indentation is possible", blockParsesTo "\na \n b \nc " "a \n b \nc ")
+    , ("single lines works like a string", blockParsesTo "  abc " "  abc ")
```
```suggestion
    , ("no whitespace is removed from the first line", blockParsesTo "  abc " "  abc ")
```

abooij reviewed (commented):
The GraphQL specification unfortunately leaves quite some ambiguity as to what constitutes valid block string syntax. For instance, _technically_, the backslash character `\` is a valid `SourceCharacter`, so that _technically_ the expression `"""\"""` is a syntactically valid block string. But I doubt any GraphQL software would accept it as such: they would rather see the `\"""` as an escaped end of a block string (and thus not the end of a block string). Similarly, _technically_, since `"` is a valid `SourceCharacter`, `"""""""` is a valid block string.

However, the specification is quite explicit about the following case. Consider the following block string:
```
"""\""""""
```
which is equivalent to:
```
"""
    \"""
"""
```
and should get interpreted as the string `"""`. The current version of the parser would consume the first 7 characters of the above blockstring, and return a value of `\`. The current version of the printer, when passed the value `"""`, would print the syntactically invalid block string `"""""""""`.

Indeed, as I commented above, it's not possible to write a total function `blockStringValue :: Printer a => Text -> a`, because not every string can be represented as a block string. There are two ways to resolve this: we can either
1. restrict the input, e.g. by introducing a `TextThatIsPrintableAsABlockString` newtype with appropriate smart constructors (todo: come up with a better name for this type), or
2. widen the output, e.g. by either
   1. making the printer partial using `Maybe`, i.e. `Parser a => Text -> Maybe a`, outputting `Nothing` when the given `Text` cannot be printed as a block string, or
   2. printing a non-block string as an alternative option. (This solution would have my preference.) This would mean that even some strings that have newlines in them would end up getting printed as a non-block string. Although perhaps unfortunate, this is necessary to not lose information or formatting. But it would involve somehow deciding if a string can be printed as a block string, which may be a nontrivial decision.

I like this last solution 2.ii. because it seems that it would allow to satisfy the property `parse (print x) = x`.

On this topic, I suppose we should also test that the interpretation of block strings is correct for those cases where it differs from non-block strings. More explicitly, we should test that something like:
```
"""
hello\nworld
"""
```
gets parsed as the Haskell string value `"hello\\nworld"` (i.e. truly as the string `hello\nworld`), i.e. that the `\n` does *not* get interpreted as a newline. And, similarly, that
```
"""
"
"""
```
gets interpreted as the Haskell string value `"\""` (i.e. `"`).

abooij commented:
I think it would be in the spirit of the GraphQL specification to say that a block string is closed by the first occurrence of `"""` which is *not* preceded by `\` (or nothing, for the case of the empty block string `""""""`).

codingkarthik commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +485,41 @@ between open close p = tok open *> p <* tok close
+  tripleQuotes = AT.string "\"\"\""
+
+-- whitespace
+ws :: Char -> Bool
+ws c = c == ' ' || c == '\t'
```
Can we rename this function to `isWhiteSpace`?

vitalibarozzi commented:
> The GraphQL specification unfortunately leaves quite some ambiguity as to what constitutes valid block string syntax. For instance, _technically_, the backslash character `\` is a valid `SourceCharacter`, so that _technically_ the expression `"""\"""` is a syntactically valid block string. But I doubt any GraphQL software would accept it as such: they would rather see the `\"""` as an escaped end of a block string (and thus not the end of a block string). Similarly, _technically_, since `"` is a valid `SourceCharacter`, `"""""""` is a valid block string.
> 
> However, the specification is quite explicit about the following case. Consider the following block string:
> 
> ```
> """\""""""
> ```
> 
> which is equivalent to:
> 
> ```
> """
>     \"""
> """
> ```
> 
> and should get interpreted as the string `"""`. The current version of the parser would consume the first 7 characters of the above blockstring, and return a value of `\`. The current version of the printer, when passed the value `"""`, would print the syntactically invalid block string `"""""""""`.
> 
> Indeed, as I commented above, it's not possible to write a total function `blockStringValue :: Printer a => Text -> a`, because not every string can be represented as a block string. There are two ways to resolve this: we can either
> 
>     1. restrict the input, e.g. by introducing a `TextThatIsPrintableAsABlockString` newtype with appropriate smart constructors (todo: come up with a better name for this type), or
> 
>     2. widen the output, e.g. by either
>        
>        1. making the printer partial using `Maybe`, i.e. `Parser a => Text -> Maybe a`, outputting `Nothing` when the given `Text` cannot be printed as a block string, or
>        2. printing a non-block string as an alternative option, removing the `StringType` split entirely. (This solution would have my preference.) This would mean that even some strings that have newlines in them would end up getting printed as a non-block string. Although perhaps unfortunate, this is necessary to not lose information or formatting. But it would involve somehow deciding if a string can be printed as a block string, which may be a nontrivial decision.
> 
> 
> I like this last solution 2.ii. because it seems that it would allow to satisfy the property `parse (print x) = x`.
> 
> On this topic, I suppose we should also test that the interpretation of block strings is correct for those cases where it differs from non-block strings. More explicitly, we should test that something like:
> 
> ```
> """
> hello\nworld
> """
> ```
> 
> gets parsed as the Haskell string value `"hello\\nworld"` (i.e. truly as the string `hello\nworld`), i.e. that the `\n` does _not_ get interpreted as a newline. And, similarly, that
> 
> ```
> """
> "
> """
> ```
> 
> gets interpreted as the Haskell string value `"\""` (i.e. `"`).

You brought some great questions here, @abooij, we are considering the possibilities, but maybe you could help us? If you are available for a quick call with us, let me know. :)

vitalibarozzi commented:
Note to self: These are important, don't forget.

On this topic, I suppose we should also test that the interpretation of block strings is correct for those cases where it differs from non-block strings. More explicitly, we should test that something like:

"""
hello\nworld
"""

gets parsed as the Haskell string value "hello\\nworld" (i.e. truly as the string hello\nworld), i.e. that the \n does not get interpreted as a newline. And, similarly, that

"""
"
"""

lorenzo commented:
@abooij Thanks for your support and the great reviews :)

I have pushed some changes that address those bugs and it feels to me like we are ready for a final round of reviews

abooij commented on test/BlockStrings.hs:
```diff
@@ -0,0 +1,96 @@
+      footnote ("Should have failed for: " <> T.unpack (tripleQuoted unparsed))
+      failure
+
+-- | We use this to wrap the first argument with I've been
+-- calling "triple-quotes", which are the delimiters for the
+-- block strings.
+blockParsesTo :: T.Text -> T.Text -> Property
+blockParsesTo unparsed expected = withTests 1 $ property $ do
+  let result =
+        case parseOnly blockString (tripleQuoted unparsed) of
+          Left l  -> Left ("Block parser failed: " <> T.pack l)
+          Right r -> Right r
+  either onError (expected ===) result
+  where
+    onError errorMsg = do
+      footnote (T.unpack errorMsg)
+      failure
+    tripleQuoted t = "\"\"\"" <> t <> "\"\"\""
```
```suggestion
-- | We use this to wrap the first argument with I've been
-- calling "triple-quotes", which are the delimiters for the
-- block strings.
blockParsesTo :: T.Text -> T.Text -> Property
blockParsesTo unparsed expected = withTests 1 $ property $ do
  let result =
        case runParser blockString (tripleQuoted unparsed) of
          Left l  -> Left ("Block parser failed: " <> l)
          Right r -> Right r
  either onError (expected ===) result
  where
    onError errorMsg = do
      footnote (T.unpack errorMsg)
      failure
    tripleQuoted t = "\"\"\"" <> t <> "\"\"\""
```

abooij commented on test/BlockStrings.hs:
```diff
@@ -0,0 +1,96 @@
+    , ("when it has non-printable chars it prints as regular string", printsAsString (VString "\NUL\n    indented!"))
+    ]
+
+-- | We use this function to tests cases that we know should
+-- fail, when we pass a function to construct wrapped the
+-- body in a delimiter, where we will probably be testing
+-- for errors using it.
+blockParseFail :: (T.Text -> T.Text) -> T.Text -> Property
+blockParseFail tripleQuoted unparsed = withTests 1 $ property $ do
+  case parseOnly blockString (tripleQuoted unparsed) of
+    Left _ -> success
+    Right _ -> do
+      footnote ("Should have failed for: " <> T.unpack (tripleQuoted unparsed))
+      failure
```
```suggestion
-- | We use this function to tests cases that we know should
-- fail, when we pass a function to construct wrapped the
-- body in a delimiter, where we will probably be testing
-- for errors using it.
blockParseFail :: (T.Text -> T.Text) -> T.Text -> Property
blockParseFail tripleQuoted unparsed = withTests 1 $ property $ do
  case runParser blockString (tripleQuoted unparsed) of
    Left _ -> success
    Right _ -> do
      footnote ("Should have failed for: " <> T.unpack (tripleQuoted unparsed))
      failure
```

abooij commented on test/BlockStrings.hs:
```diff
@@ -0,0 +1,96 @@
+
+    -- Printing
+    , ("a line with indentation prints as blockstring", printsAsBlockString (VString "foo\n    indented!"))
+    , ("when starts with indentation prints a regular string", printsAsString (VString "    indented!"))
```
I'm not sure what the purpose of this test is. This particular string can *technically* be printed as a block string, without loss of semantics:
```
"""    indented!"""
```
It's just that the particular print of the current code would not preserve semantics. But that's a strange thing to test for.

lorenzo replied:
This is obviously erring on the safe side. It checks the heuristic "should print as block sting if there is at least one line with 0 indentation"

abooij commented on test/BlockStrings.hs:
```diff
@@ -0,0 +1,96 @@
+    -- Printing
+    , ("a line with indentation prints as blockstring", printsAsBlockString (VString "foo\n    indented!"))
+    , ("when starts with indentation prints a regular string", printsAsString (VString "    indented!"))
+    , ("when has escaped newlines prints as regular string", printsAsString (VString "foo\\n    indented!"))
```
What's wrong with the following?
```
"""
foo\n    indented!
"""
```

abooij replied:
To be honest, these `printsAsBlockString` and `printsAsString` test cases do not make me very convinced that the code is free of bugs.

lorenzo replied:
This was discussed in the call. We decided to err on the safe side when printing strings and choose to use a regular string if we detected any escaped new lines.

abooij replied:
To put it differently: a test suite should test when _any_ implementation could be considered (in)correct, rather than verifying what the expected behavior of a known implementation is.

lorenzo replied:
Got it, that makes sense to me if you want to use tests as documentation about the standard. 

What type of tests do you suggest we have to show that we are indeed printing things as block texts?

abooij replied:
I think it suffices the following two categories of tests suffice:
- roundtrip tests, namely that `parse . print = id` (the other way around is not important for us) (this is covered by `genBlockText`), and
- parsing tests, namely that a specific block string is interpreted as a specific Haskell text value (this is covered by `blockTest`).

I don't think it is crucial to test that the printer works correctly in isolation - so just remove the last 4 lines in `blockTest`.

abooij reviewed (changes requested):
@lorenzo The printing code still has bugs, along the lines I have indicated before.

Please add this test to `test/Keywords.hs`:

```
propHandleTripleQuote :: Property
propHandleTripleQuote = property $ testRoundTripValue $ VString "\"\"\""
```
and correspondingly add this test to `primitiveTests`.

I have unmarked some comments as "resolved" that I don't think have been resolved.

I'll do a thorough review when the existing comments have been fully addressed.

vitalibarozzi commented:
@abooij So hey, we addressed the problems that were discussed, but maybe in the noise we missed something again. But, FYI:
- We did a general cleanup and addressed long lost comments, specially the one involving "headline", if you remember.
- We fixed the printer printing the `"""` as if it were a block string.

abooij commented on test/BlockStrings.hs:
```diff
@@ -0,0 +1,65 @@
+    , ("empty X lines", blockParsesTo "\n\n\n\n\n\n" "")
+    , ("preserves escaped newlines", blockParsesTo "\nhello\\nworld\n" "hello\\nworld")
+    , ("", blockParsesTo "\n\"\n" "\"")
+    , ("escaped triple-quotes are ignored as block terminator", blockParsesTo "\n   \\\"\"\"hey\n   friends\n" "\\\"\"\"hey\nfriends")
```
This isn't quite right, I think. `\"""` should get parsed to `"""` - see the "Semantics" section in the [spec](http://spec.graphql.org/June2018/#sec-String-Value).
```suggestion
    , ("escaped triple-quotes are ignored as block terminator", blockParsesTo "\n   \\\"\"\"hey\n   friends\n" "\"\"\"hey\nfriends")
```

lorenzo replied:
In that case, it does not parse as a block string, but a regular string. I'll change the test so that it reads `blockParseFail` instead

lorenzo replied:
I did the change but on second thought I'm not sure this is a bug in the parser or this is the correct behavior. I'll need to study it more when I get a chance again to come back to this.

If you think that it is a bug, please let me know

abooij replied:
The following is a *valid* block string:
```
"""
   \"""hey
   friends
"""
```
which should get parsed to the value:
```
"""hey
friends
```
i.e. to the Haskell string
```
"\"\"\"hey\nfriends"
```
(no backslash anymore)

This corresponds to my suggestion for the tests above. There is a bug in the parser.

abooij commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +483,84 @@ between open close p = tok open *> p <* tok close
+          []              -> headline
+          reformatedLines -> rebuild (sanitize $ headline:reformatedLines)
+
+    -- This function is used to parse the body of the string
+    -- while counting stuff that we might need for escaping,
+    -- for example.
```
```suggestion
    -- Take characters from the block string until the first
    -- non-escaped triple quotes.
```

abooij reviewed (commented):
Thanks for addressing the comments. I think this PR is now going in a great direction.

I have done a thorough review and have left some more comments. Most of them are typos and coding style - I've tried to leave suggestions so that they can be dealt with quickly. I couldn't leave a suggestion for code indentation somewhere (about dollar characters lining up) because the github UI wouldn't let me.

There's one behavioral issue that still needs some attention. There seems to be a mistake in the property tests, where `\"""` should get parsed as `"""`. I left a suggestion to fix the test. Correspondingly, `extractText` should replace instances of `\"""` with `"""`. I'm sorry I didn't notice this issue earlier.

lorenzo commented:
@abooij thanks for all the suggestions. I have applied them all and fixed some minor errors I found in them.

Also left a question about one of the test cases you changed. Please let me know what you think of it.

vitalibarozzi commented:
Hey, @abooij, we fixed the bug you've mentioned and the corresponding test. Care to take a look? :)

codingkarthik commented on src/Language/GraphQL/Draft/Parser.hs:
```diff
@@ -479,3 +483,83 @@ between open close p = tok open *> p <* tok close
+
+    extractText = 
+      -- The reason we have this replace here is to convert
+      -- an escapped triple-quotes to the way it should be
```
```suggestion
      -- an escaped triple-quotes to the way it should be
```

codingkarthik commented:
@vitalibarozzi @lorenzo Auke and myself are now happy with this PR now and is ready to merge at-least from the code PoV. The only blocker now is to get the CI to pass the tests now. The tests are timing out due to some modifications we made to the tests recently. I'll look into it and make sure that this PR gets merged ASAP.

codingkarthik reviewed (approved):
LGTM (sorry for the late approval)

codingkarthik commented:
@vitalibarozzi @lorenzo There's a warning while building the code.

```
src/Language/GraphQL/Draft/Syntax.hs:95:1: warning: [-Wdeprecations]
    Module ‘Data.Text.Prettyprint.Doc’ is deprecated:
      Use "Prettyprinter" instead.
   |
95 | import                          Data.Text.Prettyprint.Doc      (Pretty (..))
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Can you fix it?

codingkarthik commented:
@vitalibarozzi @lorenzo There's a warning while building the code.

```
src/Language/GraphQL/Draft/Syntax.hs:95:1: warning: [-Wdeprecations]
    Module ‘Data.Text.Prettyprint.Doc’ is deprecated:
      Use "Prettyprinter" instead.
   |
95 | import                          Data.Text.Prettyprint.Doc      (Pretty (..))
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Can you fix it?

codingkarthik commented:
Also, can you merge the latest master into your branch? Hopefully that should solve the CI issues.

vitalibarozzi commented:
@codingkarthik Done. :)

lorenzo commented:
@codingkarthik @abooij  good to merge now?
